python/ray/dataframe/dataframe.py

```diff
@@ -996,24 +987,56 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
         """
         axis = pd.DataFrame()._get_axis_number(axis)
 
-        if is_list_like(func) and not all([isinstance(obj, str)
-                                           for obj in func]):
-            raise NotImplementedError(
-                "To contribute to Pandas on Ray, please visit "
-                "github.com/ray-project/ray.")
-
-        if axis == 0 and is_list_like(func):
-            return self.aggregate(func, axis, *args, **kwds)
         if isinstance(func, compat.string_types):
             if axis == 1:
                 kwds['axis'] = axis
             return getattr(self, func)(*args, **kwds)
+        elif isinstance(func, dict):
+            if axis == 1:
+                raise TypeError(
+                    "(\"'dict' object is not callable\", "
+                    "'occurred at index {0}'".format(self.index[0]))
+            has_list = list in map(type, func.values())
+            part_ind_tuples = [(self._col_metadata[key], key) for key in func]
+
+            # tup[1] is the key of the dict
+            # tup[0][0] is partition index
+            # tup[0][1] is the index within the partition
+            if has_list:
+                # if input dict has a list, the function to apply must wrap
+                # single functions in lists as well to get the desired output
+                # format
+                result = [_deploy_func.remote(
+                    lambda df: df.iloc[:, tup[0][1]].apply(
+                        func[tup[1]] if is_list_like(func[tup[1]])
+                        else [func[tup[1]]]),
+                    self._col_partitions[tup[0][0]])
+                    for tup in part_ind_tuples]
```

Comment:
You can auto-unbox this by replacing `tup` with `(p, iwp), key` or something like that.